x-pack/plugins/observability/public/hooks/slo/use_create_slo.ts

```diff
@@ -5,75 +5,91 @@
  * 2.0.
  */
 
-import { useCallback, useState } from 'react';
+import { useMutation, useQueryClient } from '@tanstack/react-query';
 import type {
   CreateSLOInput,
   CreateSLOResponse,
+  FindSLOResponse,
   UpdateSLOInput,
   UpdateSLOResponse,
 } from '@kbn/slo-schema';
 
 import { useKibana } from '../../utils/kibana_react';
 
-interface UseCreateOrUpdateSlo {
-  loading: boolean;
-  success: boolean;
-  error: string | undefined;
-  createSlo: (slo: CreateSLOInput) => void;
-  updateSlo: (sloId: string, slo: UpdateSLOInput) => void;
-}
-
-export function useCreateOrUpdateSlo(): UseCreateOrUpdateSlo {
+export function useCreateOrUpdateSlo() {
   const { http } = useKibana().services;
-  const [loading, setLoading] = useState(false);
-  const [success, setSuccess] = useState(false);
-  const [error, setError] = useState<string | undefined>(undefined);
+  const queryClient = useQueryClient();
 
-  const createSlo = useCallback(
-    async (slo: CreateSLOInput) => {
-      setLoading(true);
-      setError('');
-      setSuccess(false);
+  const createSlo = useMutation(
+    ({ slo }: { slo: CreateSLOInput }) => {
       const body = JSON.stringify(slo);
-
-      try {
-        await http.post<CreateSLOResponse>(`/api/observability/slos`, { body });
-        setSuccess(true);
-      } catch (e) {
-        setError(e);
-      } finally {
-        setSuccess(false);
-        setLoading(false);
-      }
+      return http.post<CreateSLOResponse>(`/api/observability/slos`, { body });
     },
-    [http]
+    {
+      mutationKey: ['createSlo'],
+      onSuccess: () => {
+        queryClient.invalidateQueries(['fetchSloList']);
+        queryClient.invalidateQueries(['fetchHistoricalSummary']);
+      },
+    }
   );
 
-  const updateSlo = useCallback(
-    async (sloId: string, slo: UpdateSLOInput) => {
-      setLoading(true);
-      setError('');
-      setSuccess(false);
+  const cloneSlo = useMutation<
+    CreateSLOResponse,
+    string,
+    { slo: CreateSLOInput; idToCopyFrom?: string },
+    { previousSloList: FindSLOResponse | undefined }
+  >(
+    ['cloneSlo'],
+    ({ slo }: { slo: CreateSLOInput; idToCopyFrom?: string }) => {
       const body = JSON.stringify(slo);
+      return http.post<CreateSLOResponse>(`/api/observability/slos`, { body });
+    },
+    {
+      onMutate: async ({ slo, idToCopyFrom }) => {
```

Comment:
This is the optimistic update to immediately show a cloned result without waiting for the API to come back with the result. If we would that the `slo` value here and use it to do `queryClient.setQueryData`, it would be the summary version, leading to the app to break. So we cheat a little, and take the id to copy the existing SLO that's inside the query cache, and use that.